//--------------------------------------------------------
//
//  L2Diagnostic
//  Author: Mark Stegall
//  Module: DiagnosticsDock.h
//
//  Purpose:
//  Determine correct operation of the Unitreee L2 Lidar hardware
//  and software.  Establish platform independent software protocols
//  for using the L2 Lidar with its Ethernet interface.
//
//  Background:
//  Unitree provides undoucmented software files in the form:
//      include files
//      example application files
//      .a Archive Library
//
//  The source files rely on an Archive library using POSIX I/O
//  No source exists for the archive Library making it diffcult
//  to debug or port usage of the L2 Lidar for other platforms.
//  The hardware has 2 mutually exclusive communication interfaces:
//      Ethernet using UDP
//      Serial UART
//  The serial UART is limited in speed and does not operate at
//  the full sensor speed of 64K/sec sample points.
//
//  Solution:
//  This software skeleton was created using directed ChatGPT AI
//  conversation targetting a QT Creator development platform.
//  It reads UPD packets from the L2, caterorizes them, performs
//  error detection for bad packets (lost), display subsample
//  of packets and optionally saves them to a CSV file.
//
//  V.02.5  2026-01-10  added Calibration and internal State dialog
//
//--------------------------------------------------------

//--------------------------------------------------------
// This uses the following Unitree L2 sources modules:
//      unitree_lidar_protocol.h
//      unitree_lidar_utilities
// The orignal source can be found at:
//      https://github.com/unitreerobotics/unilidar_sdk2
//      under License: BSD 3-Clause License (see files)
//
// Corrections/additions have been made to these 2 files
//--------------------------------------------------------

//--------------------------------------------------------
// GPL-3.0 license
//
// This file is part of L2diagnsotic.
//
// L2diagnsotic is free software : you can redistribute it and /or modify it under
// the terms of the GNU General Public License as published by the Free Software Foundation,
// either version 3 of the License, or (at your option) any later version.
//
// L2diagnsotic is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU General Public License for more details.
// You should have received a copy of the GNU General Public License along with L2diagnsotic.
// If not, see < https://www.gnu.org/licenses/>.
//--------------------------------------------------------

#pragma once

#include <QDockWidget>

QT_BEGIN_NAMESPACE
namespace Ui {
class DiagnosticsDock;
}
QT_END_NAMESPACE

// needed device data structures
#include "unitree_lidar_protocolL2.h"

//--------------------------------------------------------
// DiagnosticsDock class definition
//
// This is a dockable window
// ui variables autogenerated from DiagnosticsDock.ui
//
//--------------------------------------------------------

//--------------------------------------------------------
//
// Usage in MainWindow ui
//
// m_diagnosticsDock = new DiagnosticsDock(this);
// addDockWidget(Qt::RightDockWidgetArea, m_diagnosticsDock);
// m_diagnosticsDock->hide();   // hide window
// m_diagnosticsDock->show();   // show window
//
//--------------------------------------------------------

class DiagnosticsDock : public QDockWidget
{
    Q_OBJECT

public:
    explicit DiagnosticsDock(QWidget* parent = nullptr);
    ~DiagnosticsDock();

    // use these calls to update this dock window
    void updateDiagnostics(const LidarInsideState& State, const LidarCalibParam& Calib,
                float range_min, float range_max );

    void updateVersion(const LidarVersionData& version);

private:
    Ui::DiagnosticsDock* ui;
};
